### ISSUE TITLE
Mempool: add support for transactions that aren't valid _yet_, but will be later

### ISSUE THREAD

**Issue by edsko** [input-output-hk/ouroboros-network#1290]

**Comment by edsko** [comment]
@dcoutts , can you clarify when exactly this will become required? I remember that such transactions will become a thing once we have Plutus integration, but I don't know if we need it before then?

**Comment by dcoutts** [comment]
Correct, this is only relevant when we integrate the ledger changes needed for Plutus.

Shelley transactions have a _required_ "not valid after slot N", and with the ledger changes with Plutus will add an _optional_ "not valid before slot M".

**Comment by edsko** [comment]
Note: we should implement #1298 before this, otherwise we might unnecessarily reject transactions.

**Comment by edsko** [comment]
Basic idea:

The mempool currently assumes that all transactions in the mempool end up in the same block, which will be forged at some slot `s`. Current `s` is the successor of the chain's current tip, although we could be a bit smarter there (https://github.com/input-output-hk/ouroboros-network/issues/1298). Adding support for transactions that aren't valid yet, but will be in the future, effectively means keeping more than one pool, for different slots `s`. One for the "next" slot `s`, just like we do now, but then additional pools at larger intervals. At one extreme, we could have one pool for every set of transactions that are valid from point `i` forward, although in most cases it will  possible to collapse two pools at slots `i` and `i+1` to the later slot `i+1` (provided hat the transactions are not yet _expired_ at slot `i+1`). For every one of these pools the ledger should be ticked to that slot (possibly forecasted? depends on what we want the semantics to be exactly.). 

This then also smoothly handles the transition around a hardfork, because if a Shelley transaction comes in just before the transition, it will end up in a pool _after_ the hard fork, and hence ticking the state to that slot will indeed have transitioned the ledger state. 

This design requires that we _know_ the slot at which a transaction becomes valid. For Plutus/MC transactions I think this is part of the transaction itself; for the HFC case, this will depend on knowing when the transition happens. I think it's ok to reject Shelley transactions when we don't even know yet when the Shelley transition will occur.

**Comment by edsko** [comment]
Marking as high priority as this will be required for Goguen.

**Comment by mrBliss** [comment]
Related: #1303.

**Comment by edsko** [comment]
# Extending the mempool to support tx validity ranges

The mempool must meet two requirements which impose somewhat conflicting
requirements on its design:

1. Adding new transactions to the mempool should not have cost that is
   proportional in the size of the mempool.
2. It should not be possible (or at least, expensive) to flood the mempool
   with transactions (protection against DoS attacks).

Not validating transactions before adding them into the mempool would satisfy
(1) but not (2). Validating transactions _in the order that they arrive_
satisfies (2) as well as (1), at the cost of imposing a rigid sequential
ordering.

That sequential ordering isn't a huge deal currently, but it becomes a bigger
problem once we start introducing transactions that are only valid _after_ a 
certain slot, especially (Plutus) scripts. 

In principle the ledger state can be updated by applying a transaction without
actually running the script. If the UTxO required by the script are removed
from the ledger state, then this protects against DoS (requirement 2) in the
same way as for regular transactions. There is however a fundamental
difference: for regular transactions we can _check_ that the transaction has the
right to spend those UTxO. For script inputs we cannot (unless we run the
script); this means that a script input that will turn out to eventually be
invalid can be used to "lock" UTxO until the time comes that the lower bound on
the transaction will be met and we _must_ run the transaction when including
it in a block.

If we _do_ run the script, we must first of all be careful: running a script is
costly, and should be done only once (marking it as valid or invalid). This is
fundamental to the design of scripts: even invalid (script) txs will be included
in blocks; they don't produce outputs but do consume their fee/gas inputs,
because the node did have the cost of evaluating them. 

The problem is that it's hard to satisfy (1):

* Scripts have validity bounds on them ("this script can only be executed
  between slots `x` and `y`"). These bounds allow the script to make certain
  assumptions about the environment in which are they run; if we do run the
  script, these assumptions should be met.
* Although we can "tick" the ledger state to any slot, this makes it difficult
  to stick to a single linear list of transactions. Once we include a
  transaction that is valid some time in the future, we would be unable to
  include any transaction that is only valid _before_ that point (we can limit
  the impact of that somewhat by limiting quite how far into the future we
  accept for lower bounds). 
* We could maintain multiple mempools, ticked to different slots, but if those
  are chained (the initial ledger state of the next pool being determined by 
  ticking the final ledger state of the previous) we'd violate (1) again,
  as inserting a transaction into an earlier pool would mean having to 
  re-evaluate the transactions in the other pools.
* If we maintain multiple mempools ticked to different slots all starting from
  the _same_ initial ledger state (rather than chaining them) we get no double
  spending protection across mempools, and hence we open ourselves up to a
  _different_ DoS attack (requirement 2; we can flood the mempool by spamming it
  with  transactions that all spend the same UTxO but are valid in different
  periods).

## Possible solution?

For ordinary transactions we consider the protection against double spending 
to be a sufficient protection against DoS as well, irrespective of the size
of the transaction's fee. Transactions that spend script outputs _must have
regular inputs as well_ to cover the fee and the gas. If we "spend" those
when adding the transaction into the mempool (without running the script)
we guard against double spending and hence against DoS. If we do _not_ "spend"
the _additional_ inputs required by the script, we avoid the locking problem
described above. Since we do not need to run the script (until we try to
produce a block), we don't have to worry about multiple mempools.

Then when we _do_ produce a block, we traverse the mempool linearly, 
re-evaluating all transactions in order (we do this now anyway, since we start
with a potentially different ledger state), excluding any transactions that are
not yet valid (leaving those in the mempool). After we have done this, the
mempool could be reordered if that would be beneficial, since we are 
re-evaluating anyway.  

Incidentally, this also prevents another potential DoS attack: since script
outputs cannot be used for gas/fees, "daily-chaining" transactions (where the
outputs of one script are used as the input to another) is not possible.

**Comment by edsko** [comment]
Ok, turns out, it's not quite this complicated. The solution above can work, _however_, it might also be possible after all to execute scripts "eagerly" (when they are added into the mempool). This is a misunderstanding on my part of how scripts work. My summary:

Basically the transaction asserts certain facts about the ledger state. Those facts are then visible to the script, but _only_ those facts. This means that we could run the script, giving it those facts, but then _verify_ those facts against the actual ledger state only once you get to the slot in which the tx is meant to be valid. This means that there is no need to tick the ledger state (this is the "fact checking part" which we can do later), and hence we can go back to a simple linear list of transactions.

**Comment by edsko** [comment]
... but it now looks like the script folks might not need this after all, and we can just reject transactions that aren't yet valid.

**Comment by edsko** [comment]
Another benefit of only running scripts when they are included in a block is that relays and other nodes that never produce blocks also don't have to execute scripts.

**Comment by mrBliss** [comment]
We should do #2591 first, it will make this issue easier to implement and test.

**Comment by mrBliss** [comment]
It turns out this is not strictly needed for Goguen.

**Comment by nfrisby** [comment]
Closing as WontFix.
